File: src/test/java/blog/cirkle/app/RestApiSimulation.java
package blog.cirkle.app;

import static io.gatling.javaapi.core.CoreDsl.*;
import static io.gatling.javaapi.http.HttpDsl.*;

import io.gatling.javaapi.core.*;
import io.gatling.javaapi.http.*;

public class RestApiSimulation extends Simulation {

	HttpProtocolBuilder httpProtocol = http.baseUrl("http://localhost:8080") // Change if needed
			.acceptHeader("application/json")
	// .bearerAuth("YOUR_TOKEN_HERE")
	; // Or set dynamically

	// Example of a GET request to retrieve a specific post
	ChainBuilder getPostById = exec(
			http("GetPostById").get("/api/v1/posts/123e4567-e89b-12d3-a456-426614174000").check(status().is(200)));

	// Scenario that exercises multiple REST endpoints
	ScenarioBuilder scn = scenario("REST_Api_Scenario").exec(getPostById).pause(1);

	{
		// We define injection steps for concurrency:
		// - rampUsers(50) over 10s, hold for 20s
		// - rampUsers(100-50=50 more, total 100) over 10s, hold for 20s
		// - rampUsers(100 more to 200)...
		// - etc., up to 1000
		// You can adapt intervals as needed
		setUp(scn.injectOpen(rampUsers(50).during(10), // Ramp to 50 in 10s
				constantUsersPerSec(50).during(20),

				rampUsers(50).during(10), // Ramp to 100 in next 10s
				constantUsersPerSec(100).during(20),

				rampUsers(100).during(10), // Ramp to 200
				constantUsersPerSec(200).during(20),

				rampUsers(300).during(10), // Ramp to 500
				constantUsersPerSec(500).during(20),

				rampUsers(500).during(10), // Ramp to 1000
				constantUsersPerSec(1000).during(20))).protocols(httpProtocol);
	}
}

File: src/gatling/scala/blog/cirkle/app/RestApiSimulation.scala
package blog.cirkle.app

import io.gatling.core.Predef._
import io.gatling.http.Predef._
import org.slf4j.LoggerFactory

class RestApiSimulation extends Simulation {

  private val logger = LoggerFactory.getLogger(classOf[RestApiSimulation])

  val httpProtocol = http
    .baseUrl("http://localhost:8080")
    .acceptHeader("application/json")
    //.bearerAuth("YOUR_TOKEN_HERE")  // adjust as needed
    .basicAuth("carlos.martinez42@gmail.com","stringcarlos.martinez42@gmail.com");

  // An example GET request chain
  val getPostById = exec { session =>
    logger.info("About to GET a specific post by ID...")
    session
  }
    .exec(
      http("GetPostById")
        .get("/api/v1/posts/123e4567-e89b-12d3-a456-426614174000")
        .check(status.is(200))
        .check(bodyString.saveAs("postResponse"))
    )
    .exec { session =>
      val resp = session("postResponse").asOption[String].getOrElse("No body")
      logger.info(s"Received GET response: $resp")
      session
    }

  val scn = scenario("REST_Api_Scenario")
    .exec(getPostById)
    .pause(1)

  setUp(
    scn.inject(
      rampUsers(50).during(10),
      constantUsersPerSec(50).during(20),
      rampUsers(50).during(10),
      constantUsersPerSec(100).during(20),
      rampUsers(100).during(10),
      constantUsersPerSec(200).during(20),
      rampUsers(300).during(10),
      constantUsersPerSec(500).during(20),
      rampUsers(500).during(10),
      constantUsersPerSec(1000).during(20)
    )
  ).protocols(httpProtocol)
}
